upb/json/printer.c

```diff
@@ -920,6 +920,53 @@ static bool printer_endmsg_noframe(
   return true;
 }
 
+static void *scalar_startstr_onlykey(
+    void *closure, const void *handler_data, size_t size_hint) {
+  upb_json_printer *p = closure;
+  UPB_UNUSED(size_hint);
+  CHK(putkey(closure, handler_data));
+  return p;
+}
+
+/* Set up handlers for an Any submessage. */
+void printer_sethandlers_any(const void *closure, upb_handlers *h) {
+  const upb_msgdef *md = upb_handlers_msgdef(h);
+
+  const upb_fielddef* type_field = upb_msgdef_itof(md, UPB_ANY_TYPE);
+  const upb_fielddef* value_field = upb_msgdef_itof(md, UPB_ANY_VALUE);
+
+  upb_handlerattr empty_attr = UPB_HANDLERATTR_INITIALIZER;
+
+  /* type_url's json name is "@type" */
+  upb_handlerattr type_name_attr = UPB_HANDLERATTR_INITIALIZER;
+  upb_handlerattr value_name_attr = UPB_HANDLERATTR_INITIALIZER;
+  strpc *type_url_json_name = upb_gmalloc(sizeof(*type_url_json_name));
+  strpc *value_json_name = upb_gmalloc(sizeof(*type_url_json_name));
+
+  type_url_json_name->ptr = upb_gstrdup("@type");
+  type_url_json_name->len = strlen(type_url_json_name->ptr);
+  upb_handlers_addcleanup(h, type_url_json_name, freestrpc);
+  upb_handlerattr_sethandlerdata(&type_name_attr, type_url_json_name);
+
+  value_json_name->ptr = upb_gstrdup("value");
+  value_json_name->len = strlen(value_json_name->ptr);
+  upb_handlers_addcleanup(h, value_json_name, freestrpc);
+  upb_handlerattr_sethandlerdata(&value_name_attr, value_json_name);
+
+  /* Set up handlers. */
+  upb_handlers_setstartmsg(h, printer_startmsg, &empty_attr);
+  upb_handlers_setendmsg(h, printer_endmsg, &empty_attr);
+
+  upb_handlers_setstartstr(h, type_field, scalar_startstr, &type_name_attr);
+  upb_handlers_setstring(h, type_field, scalar_str, &empty_attr);
+  upb_handlers_setendstr(h, type_field, scalar_endstr, &empty_attr);
+
+  upb_handlers_setstartstr(h, value_field, scalar_startstr_onlykey,
```

Comment:
Please add a comment that this is not the full and correct JSON encoding for Any, and that it requires further processing by the client based on the type URL.

Reply:
done